src/main/java/io/reactivex/rxjava3/internal/operators/observable/ObservableBufferTimed.java

```diff
@@ -0,0 +1,564 @@
+                queue.offer(b);
+                done = true;
+                if (enter()) {
+                    QueueDrainHelper.drainLoop(queue, downstream, false, null, this);
+                }
+            }
+            DisposableHelper.dispose(timer);
+        }
+
+        @Override
+        public void dispose() {
+            DisposableHelper.dispose(timer);
+            upstream.dispose();
+        }
+
+        @Override
+        public boolean isDisposed() {
+            return timer.get() == DisposableHelper.DISPOSED;
+        }
+
+        @Override
+        public void run() {
+            U next;
+
+            try {
+                next = Objects.requireNonNull(bufferSupplier.get(), "The bufferSupplier returned a null buffer");
+            } catch (Throwable e) {
+                Exceptions.throwIfFatal(e);
+                downstream.onError(e);
+                dispose();
+                return;
+            }
+
+            U current;
+
+            synchronized (this) {
+                current = buffer;
+                if (current != null) {
+                    buffer = next;
+                }
+            }
+
+            if (current == null) {
+                DisposableHelper.dispose(timer);
+                return;
+            }
+
+            fastPathEmit(current, false, this);
+        }
+
+        @Override
+        public void accept(Observer<? super U> a, U v) {
+            downstream.onNext(v);
+        }
+    }
+
+    static final class BufferSkipBoundedObserver<T, U extends Collection<? super T>>
+    extends QueueDrainObserver<T, U, U> implements Runnable, Disposable {
+        final Supplier<U> bufferSupplier;
+        final long timespan;
+        final long timeskip;
+        final TimeUnit unit;
+        final Worker w;
+        final List<U> buffers;
+
+        Disposable upstream;
+
+        BufferSkipBoundedObserver(Observer<? super U> actual,
+                Supplier<U> bufferSupplier, long timespan,
+                long timeskip, TimeUnit unit, Worker w) {
+            super(actual, new MpscLinkedQueue<>());
+            this.bufferSupplier = bufferSupplier;
+            this.timespan = timespan;
+            this.timeskip = timeskip;
+            this.unit = unit;
+            this.w = w;
+            this.buffers = new LinkedList<>();
+        }
+
+        @Override
+        public void onSubscribe(Disposable d) {
+            if (DisposableHelper.validate(this.upstream, d)) {
+                this.upstream = d;
+
+                final U b; // NOPMD
+
+                try {
+                    b = Objects.requireNonNull(bufferSupplier.get(), "The buffer supplied is null");
+                } catch (Throwable e) {
+                    Exceptions.throwIfFatal(e);
+                    d.dispose();
+                    EmptyDisposable.error(e, downstream);
+                    w.dispose();
+                    return;
+                }
+
+                buffers.add(b);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 This line may not be thread-safe: **257**. At this line the following field is not protected by synchronization: `buffers`. Your code typically protects this field with synchronization on object: `this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `this`. For example, like at lines: **380**, **363**, **318**, **289**, **270**, and **342**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **270**, **289**, **318**, **342**, **363**, and **380**.